### PR TITLE
KS: Disabled Icon

### PR DESCRIPTION
**Pull request by tfamula**
Feature Wiki:
[KS: Disabled Icon](https://www.ilias.de/docu/goto_docu_wiki_wpage_4918_1357.html)
[Presentation of Items WITHOUT Read-Permission](https://www.ilias.de/docu/goto.php?target=wiki_1357_Presentation_of_Items_WITHOUT_Read-Permission)
[Improve Usability of List of Portfolios](https://www.ilias.de/docu/goto.php?target=wiki_1357_Improve_Usability_of_List_of_Portfolios)

### CODE REVIEW

File: src/UI/Factory.php [hunk 1]
```diff
@@ -22,10 +22,14 @@
 	 *     They indicate the functionality or nature of a text-element or context:
 	 *     Icons will mainly be used in front of object-titles, e.g. in the
 	 *     header, the tree and in repository listing.
+	 *     Icons can be disabled. Disabled Icons visually communicate that they
+	 *     can not be worked with by users, i.e. a precondition is not met yet or
+	 *     status is offline. Technically this means that read access is not granted.
```
Comment (klees) [inline src/UI/Factory.php]: Please remove that `Technically...` part and just add that as another example why an icon possibly cannot be used.

File: src/UI/Factory.php [hunk 2]
```diff
@@ -22,10 +22,14 @@
 	 *     They indicate the functionality or nature of a text-element or context:
 	 *     Icons will mainly be used in front of object-titles, e.g. in the
 	 *     header, the tree and in repository listing.
+	 *     Icons can be disabled. Disabled Icons visually communicate that they
+	 *     can not be worked with by users, i.e. a precondition is not met yet or
```
Comment (klees) [inline src/UI/Factory.php]: I do not think, that the "icons" are the things that the user can work with. It rather is the indicated funtionality that the user cannot use.

File: src/UI/Factory.php [hunk 3]
```diff
@@ -40,11 +44,16 @@
 	 *     1: Icons MUST be used to represent objects or context.
 	 *     2: Icons MUST be used in combination with a title or label.
 	 *     3: An unique Icon MUST always refer to the same thing.
+	 *     4: Preconditioned objects whose precondition is not met yet
+	 *        feature Disabled Icons
+	 *     5: Objects to which a user has no read access features Disabled Icons.
```
Comment (klees) [inline src/UI/Factory.php]: These imo should go to the `context`-field (have a look [here](https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-3/src/UI/docu/rules.md#interfaces-to-factories)).

File: src/UI/Factory.php [hunk 4]
```diff
@@ -40,11 +44,16 @@
...
 	 *   style:
 	 *     1: Icons MUST have a class indicating their usage.
 	 *     2: Icons MUST be tagged with a CSS-class indicating their size.
+	 *     3: Disabled Icons will be shaded grey.
```
Comment (klees) [inline src/UI/Factory.php]: This is no rule (not MUST, SHOULD, ...), also composition already states something similar.

**Review by klees — COMMENTED** [review]
Hi @tfamula-ilias, thx for the contribution. This LGTM in general. I left some remarks on your new text snippets, feel free to ask for further feedback.

**Review by klees — CHANGES_REQUESTED** [review]
Forgot to click on "Request changes"...

File: src/UI/Factory.php (hunk 2 above)
Comment (Amstutz) [reply src/UI/Factory.php]: Maybe rephrase to: "... communicate that the depicted functionality is not available to the user."

File: src/UI/Factory.php (hunk 1 above)
Comment (Amstutz) [reply src/UI/Factory.php]: Should they even be displayed in the given example? Can you elaborate a little on this?

File: src/UI/Factory.php [hunk 5]
```diff
@@ -40,11 +44,16 @@
 	 *     1: Icons MUST be used to represent objects or context.
 	 *     2: Icons MUST be used in combination with a title or label.
 	 *     3: An unique Icon MUST always refer to the same thing.
+	 *     4: Preconditioned objects whose precondition is not met yet
+	 *        feature Disabled Icons
```
Comment (Amstutz) [inline src/UI/Factory.php]: Like visible, but not read?

File: src/UI/Factory.php [hunk 6]
```diff
@@ -40,11 +44,16 @@
...
+	 *     5: Objects to which a user has no read access features Disabled Icons.
 	 *   style:
 	 *     1: Icons MUST have a class indicating their usage.
 	 *     2: Icons MUST be tagged with a CSS-class indicating their size.
+	 *     3: Disabled Icons will be shaded grey.
 	 *   accessibility:
 	 *     1: Icons MUST use aria-label.
+	 *     2: Disabled Icons MUST bear an aria-label indicating the special status.
```
Comment (Amstutz) [inline src/UI/Factory.php]: +1, good one.

**Review by Amstutz — CHANGES_REQUESTED** [review]
hi @tfamula-ilias ,

Thanks for the FR. In general I like it and I believe it to be useful. However, can you a bit more specific about the context in which this is intended to be used? I guess it is something like, object is visible but not readable. Is there an FR asking for this? If so, link it in the description of this entry. Also make the example in the purpose statement a bit more specific.

File: src/UI/Factory.php (hunk 1 above)
Comment (klees) [reply src/UI/Factory.php]: ` i.e. a precondition is not met yet, status is offline or read access is not granted.`

Comment (Amstutz) [reply src/UI/Factory.php]: @klees ?

Probably my point was not clear enough:
Please be more specific in which case the disabled icon should be displayed and in which case the object is not displayed at all. I can only guess, that this is intended to be used if somebody has read access but others might not, correct? E.g. some admin would see this and this indicates that others might not (as the "offline" label currently does), correct? In this case others would not see the object at all, and the admin gets a hint of the state of this object. Or is it, somebody has "visible" rights but not "read"? Or both?

This differentiation seems important to me, since the first indicates the status of something for the user configuring it, but not the state of access for the current user itself, while the second directly communicates the options for the user currently staring at the icon.

Comment (alex40724) [reply src/UI/Factory.php]: objects and their icons are usually shown, if visible permission is given. For these visible objects the disabled icon should indicate if "read access" is given or not (not taking any other permissions into account). Read access may not be given if:
- no read permission is given
- no precondition is met
- object is offline
- current date/time is not in activation period
- ... (anything that limits read access in objects access classes)

Comment (klees) [reply src/UI/Factory.php]: Ok, I thought your remark was for me. I think that my opinion here is exactly opposite. I would not want to discuss, solve or document these questions in the context of the UI-framework. For me `Disabled Icons visually communicate that they can not be worked with by users` with some examples is absolutely enough.

Comment (alex40724) [reply src/UI/Factory.php]: "Disabled Icons visually communicate that they can not be worked with by users" is ok for me too. This was just a technical explanation. We add links to the concrete FRs.

Comment (Amstutz) [reply src/UI/Factory.php]: Hmm, somehow i still believe that my point is not yet understood.

> For me Disabled Icons visually communicate that they can not be worked with by users with some examples is absolutely enough.

@klees I disagree. 

>Disabled Icons visually communicate that they can not be worked with by users

You see that there are two different options into which this could be understood? It could mean that there users out there which have no access (cause this is offline or similar), and this is shown to someone that might have all possible permissions. Or this can always ONLY directly relate to the permission of the user staring at it. From what I gather from @alex40724  this is the case. If we introduce state on icons, we need to agree on the meaning of those states. 

For me some wording as:
"The disabled is always to be understood in the context of the current user and not some global state such as offline. E.g. if a user has the permission to read and write the content of an offline object, then the icon shown to this user for this object should not be marked as disabled."

would be helpful.

However, if it is clear to anybody else, I can live without.

Comment (klees) [reply src/UI/Factory.php]: > You see that there are two different options into which this could be understood?

Yes, now I see =) Your proposed wording is fine for me. I'm 100% sure if that distinction is really required, though.

File: src/UI/Factory.php (hunk 3 above)
Comment (tfamula) [reply src/UI/Factory.php]: @klees Could you please explain why the context field is never shown in the KS documentation, although it exists in the code? I attached two screenshots to show you what I mean. As you can see, also the other fields are missing, e.g. background in this example. Only the fields description and rules are shown in the documentation.

ks_context
ks_context_missing

Comment (Amstutz) [reply src/UI/Factory.php]: @tfamula-ilias you are right, they are missing. Please create a mantis ticket. This is a bug, which @klees has nothing to do with (system styles are manged by me). You can directly assign this to me. Thx.

File: src/UI/Factory.php (hunk 1 above)
Comment (alex40724) [reply src/UI/Factory.php]: Oliver is proposing to show the disabled icon for users having write permission (authors), if a portfolio is not shared/offline:
https://www.ilias.de/docu/goto_docu_wiki_wpage_5052_1357.html

Comment (Amstutz) [reply src/UI/Factory.php]: @alex40724 Thanks a lot. Now we are at the core of the question: Is this fine? Can the disabled icon be used both ways: To mark something as not accessible for the current user AND/OR to mark something as not accessible for other users? I could live with that, but would like it to be clearly pointed out somehow.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 04 JUN 2018 : We highly appreciate this PR and accept it with the following changes / clarifications: 
(1) Disabled Icons visually communicate that the depicted functionality is not available for the intended audience.
(2) Additionally, we prefer a pure CSS solution for showing the status of a disabled icon (even if not supported by IE). This will make it easy to support plugins as well as customized skins.

**Comment by alex40724** [comment]
@Amstutz We used the context yaml field now, however get an error when running the test cases:

MainFactoryTest::test_kitchensink_info_context with data set #0 (ReflectionMethod Object (...), 'icon')
TODO (icon): remove 'context' field, method returns a factory.
Failed asserting that an array does not have the key 'context'.

/htdocs/ilias/tests/UI/AbstractFactoryTest.php:296

**Comment by Amstutz** [comment]
@alex40724 It's pretty much what is stated in the message, icon returns a factory (depicts a family of UI-Components) and therefore can not have a context. You have the following options:

1. Do some raphrasing and move the description of icons with the disabled attribut to usage, e.g.:

`Icons can be disabled. Disabled Icons visually communicate that the depicted functionality is not available for the intended audience. Preconditioned objects whose precondition is not met yet therefore feature Disabled Icons as well as bbjects to which a user has no read access.`

2. Propose to drop this condition for factories such as icons. We defined the context as follows:

> The context states: where this control is used specifically (this list might not be complete) and how common this control is used

We thought about specific instances, one could argue, that this could be extensible to complete families with somewhat more generic context descriptions. However in this case, I think it fits really better in usage then in context.

**Comment by alex40724** [comment]
@amstutz Personally I do not understand why families of components should not have a context field. @klees suggested to put this information into the "context" field, see https://github.com/ILIAS-eLearning/ILIAS/pull/1073#discussion_r191352427 so I am not sure if moving this to "usage" is ok for him.

Could both of you discuss this and decide on this please? Thanks.

**Comment by klees** [comment]
To me this is a valid usage of the `context` section. We only have a `usage` section in `rules`, the text would not really fit in there (it's no rule). I general think that the separation of `purpose` and `context` is valuable, as `purpose` should describe the general scenario, while `context` describes the concrete locations. I would propose to just rename `context` to `usage` and remove that rule that forbids the usage of that field here. IMO we did not think of the requirement, that properties of components need to be described.

**Comment by Amstutz** [comment]
@klees, I would not rename the context to usage atm. We should not change the template too often. Please just remove the check that families can not have a context, since you introduced it. I am completely fine with that.

@alex40724 : No change required from your part. We remove the failing check. Thx., for the notice about the issue.

**Comment by klees** [comment]
@Amstutz It's not just the check, it's the rule that needs to be removed.  This should go to JF imo. The check than can follow.

**Comment by alex40724** [comment]
Can I merge this somehow without an additional JF iteration (add these two lines later)?

**Comment by Amstutz** [comment]
@klees: You are right.

@alex40724 : Yes. Just remove the two lines for the moment. I will propose a separate request for the JF discussion and put your two lines in there, so you do not have to worry about this anymore.
